### PR TITLE
Alarm replies

### LINKED ISSUE #60: Market alarms recipients do not return their address
The recipient should return that info in the response in order to allow the cosmwasm reply mechanism to remove them.

### PR DESCRIPTION
**Pull request by go2bodhi**
Closes #60

### CODE REVIEW

File: contracts/lease/src/contract/state/mod.rs [hunk 1]
```diff
@@ -131,6 +128,11 @@ impl Response {
             next_state: next_state.into(),
         }
     }
+
+    pub fn attach_alarm_response(mut self, env: &Env) -> ContractResult<Self> {
+        self.cw_response = self.cw_response.set_data(to_binary(&env.contract.address)?);
+        Ok(self)
+    }
```
[the remark this answers is not in the document]
Comment (gmanev7) [reply contracts/lease/src/contract/state/mod.rs:135]: there is an fn `platform::response::response` that is a good candidate

File: contracts/lease/src/contract/state/ica_connector.rs [hunk 2]
```diff
@@ -82,7 +81,7 @@ impl<const PRECONNECTABLE: bool, Connectee> Enterable for IcaConnector<PRECONNEC
 where
     Connectee: IcaConnectee + DexConnectable,
 {
-    fn enter(&self, _deps: Deps<'_>, _env: Env) -> ContractResult<Batch> {
+    fn enter(&self, _deps: Deps<'_>, _env: &Env) -> ContractResult<Batch> {
```
[the remark this answers is not in the document]
Comment (KirilMihaylov) [reply contracts/lease/src/contract/state/ica_connector.rs:84]: Motivation is that we do **not** need an owned copy of the data, thus I think we do follow it.

The other thing is that, if we do pass the ownership, the data needs to be cloned which might in turn prove useless if an error occurs during the execution.

File: packages/platform/src/response.rs [hunk 3]
```diff
@@ -1,12 +1,31 @@
...
-    cosmwasm_ext::Response,
-    cosmwasm_std::{to_binary, StdResult},
+    cosmwasm_ext::Response as CwResponse,
+    cosmwasm_std::{to_binary, Binary, StdResult},
 };
-use serde::Serialize;
 
-pub fn response<T>(resp: &T) -> StdResult<Response>
+#[inline]
+pub fn response<T>(response: &T) -> StdResult<Response>
```
Comment (gmanev7) [inline packages/platform/src/response.rs:9]: [here and below] let's keep the `Std` prefix for the Rust Std namespace and pertaining to that space types. For uniformity let's use the `Cw` prefix.

File: contracts/admin/src/lib.rs [hunk 4]
```diff
@@ -40,10 +40,10 @@ pub fn instantiate(
 }
 
 #[cfg_attr(feature = "contract-with-bindings", entry_point)]
-pub fn migrate(deps: DepsMut<'_>, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
+pub fn migrate(deps: DepsMut<'_>, _env: Env, _msg: MigrateMsg) -> StdResult<Response> {
```
Comment (gmanev7) [inline contracts/admin/src/lib.rs:43]: what is the motivation for switching to the cosmwasm error type?

File: contracts/dispatcher/src/contract.rs [hunk 5]
```diff
@@ -75,12 +75,12 @@ pub fn instantiate(
...
 
 #[cfg_attr(feature = "contract-with-bindings", entry_point)]
-pub fn migrate(deps: DepsMut<'_>, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
+pub fn migrate(deps: DepsMut<'_>, _env: Env, _msg: MigrateMsg) -> StdResult<Response> {
     versioning::update_software(deps.storage, version!(CONTRACT_STORAGE_VERSION))?;
 
     SingleUserAccess::remove_contract_owner(deps.storage);
 
-    response::response(versioning::release()).map_err(Into::into)
+    response(versioning::release()).map(Into::into)
```
Comment (gmanev7) [inline contracts/dispatcher/src/contract.rs:83]: let's follow the function calling convention by providing the function module

File: packages/platform/src/response.rs [hunk 6]
```diff
@@ -1,12 +1,31 @@
...
 where
     T: Serialize + ?Sized,
 {
-    Ok(Response::new().set_data(to_binary(resp)?))
+    response_with_messages(CwResponse::new(), response)
+}
+
+pub fn response_with_messages<T, U>(messages: T, response: &U) -> StdResult<Response>
```
Comment (gmanev7) [inline packages/platform/src/response.rs:16]: cosmetic: change the order of fn parameters to reflect the fn name, first the response, then the messages

File: contracts/lease/src/contract/state/controller.rs [hunk 7]
```diff
@@ -130,9 +130,9 @@ pub fn execute(
                  cw_response,
                  next_state,
              }| {
-                super::save(deps.storage, &next_state)?;
-
-                Ok(cw_response)
+                super::save(deps.storage, &next_state)
+                    .map(move |()| cw_response)
+                    .map_err(Into::into)
```
Comment (gmanev7) [inline contracts/lease/src/contract/state/controller.rs:135]: two arguments against applying the functional style here:
- only this occurrence in this module changed, there are at least a few more
- the previous code was shorter and easier to read

File: contracts/lease/src/contract/state/ica_connector.rs [hunk 8]
```diff
@@ -21,7 +20,7 @@ use crate::{
 use super::{ica_post_connector::Postpone, State};
 
 pub(crate) trait Enterable {
-    fn enter(&self, deps: Deps<'_>, _env: Env) -> ContractResult<Batch>;
+    fn enter(&self, deps: Deps<'_>, _env: &Env) -> ContractResult<Batch>;
```
Comment (gmanev7) [inline contracts/lease/src/contract/state/ica_connector.rs:23]: we've agreed to postpone such a change

File: contracts/admin/src/lib.rs (hunk 4 above)
Comment (KirilMihaylov) [reply contracts/admin/src/lib.rs:43]: Conversion to our error type is not needed as of now. This change was only because of minimization.

Comment (gmanev7) [reply contracts/admin/src/lib.rs:43]: we do not want to change the interface later when we have to implement the method. As you may see, all methods at this point return StdResult<Response, ContractError>. If you wanna minimize, then define a `type ContractResult<T> = StdResult<T, ContractError>` as we did at the Lease, for example

Comment (gmanev7) [reply contracts/admin/src/lib.rs]: that applies to all occurrences, pls

Comment (gmanev7) [reply contracts/admin/src/lib.rs]: in all contracts

**Pull request merged**